File: src/core/execution/nodes/telegram_memory_creator.py
"""
Telegram Memory Creator Node
Stores Telegram messages and creates summaries per chat
"""
import time
import weakref
from typing import Dict, Any, Optional, List
from ..node_base import BaseNode, ExecutionContext
from ....utils.logger import get_logger

logger = get_logger(__name__)


class TelegramMemoryCreatorNode(BaseNode):
    """
    Stores Telegram messages and creates summaries per chat.
    
    Unlike regular MemoryCreator which stores Q&A pairs, this stores
    individual messages with metadata (user_id, username, chat_id).
    
    Tracks message count per chat_id and triggers summarization
    when threshold is reached.
    
    Inputs:
        message: Message text (required)
        user_id: Telegram user ID (required)
        username: Telegram username (optional)
        chat_id: Telegram chat/group ID (required)
        storage_instance: StorageInterface instance (required)
        model: ObeliskLLM instance (required for summarization)
    
    Properties:
        summarize_threshold: Number of messages before summarizing (default: 50)
    
    Outputs:
        success: Boolean indicating if message was stored
        message_count: Current message count for this chat
        summary_created: True if a summary was just created
    """
    
    # Class-level cache using WeakKeyDictionary to auto-cleanup when storage is GC'd
    # storage_instance -> chat_id -> count
    _message_counts: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
    
    # storage_instance -> chat_id -> [messages]
    _message_buffers: weakref.WeakKeyDictionary = weakref.WeakKeyDictionary()
    
    def __init__(self, node_id: str, node_data: Dict[str, Any]):
        """Initialize telegram memory creator node"""
        super().__init__(node_id, node_data)
        self._summarize_threshold = int(self.metadata.get('summarize_threshold', 50))
    
    def _get_message_count(self, storage_instance, chat_id: str) -> int:
        """Get message count for a chat"""
        if storage_instance not in self._message_counts:
            self._message_counts[storage_instance] = {}
        if chat_id not in self._message_counts[storage_instance]:
            self._message_counts[storage_instance][chat_id] = 0
        return self._message_counts[storage_instance][chat_id]
    
    def _increment_message_count(self, storage_instance, chat_id: str) -> int:
        """Increment and return message count for a chat"""
        if storage_instance not in self._message_counts:
            self._message_counts[storage_instance] = {}
        if chat_id not in self._message_counts[storage_instance]:
            self._message_counts[storage_instance][chat_id] = 0
        self._message_counts[storage_instance][chat_id] += 1
        return self._message_counts[storage_instance][chat_id]
    
    def _get_message_buffer(self, storage_instance, chat_id: str) -> List[Dict]:
        """Get message buffer for a chat"""
        if storage_instance not in self._message_buffers:
            self._message_buffers[storage_instance] = {}
        if chat_id not in self._message_buffers[storage_instance]:
            self._message_buffers[storage_instance][chat_id] = []
        return self._message_buffers[storage_instance][chat_id]
    
    def _add_to_buffer(self, storage_instance, chat_id: str, message_data: Dict):
        """Add message to buffer"""
        buffer = self._get_message_buffer(storage_instance, chat_id)
        buffer.append(message_data)
        # Keep buffer size reasonable (2x threshold)
        max_size = self._summarize_threshold * 2
        if len(buffer) > max_size:
            self._message_buffers[storage_instance][chat_id] = buffer[-max_size:]
    
    def _clear_buffer(self, storage_instance, chat_id: str):
        """Clear message buffer after summarization"""
        if storage_instance in self._message_buffers and chat_id in self._message_buffers[storage_instance]:
            self._message_buffers[storage_instance][chat_id] = []
    
    def _summarize_messages(
        self,
        llm,
        messages: List[Dict],
        chat_id: str
    ) -> Optional[Dict[str, Any]]:
        """
        Summarize chat messages using LLM.
        """
        if not messages:
            return None
        
        try:
            # Format messages for summarization
            conversation_text = ""
            for msg in messages:
                username = msg.get('username') or msg.get('user_id', 'Unknown')
                text = msg.get('message', '')
                timestamp = msg.get('timestamp', '')
                conversation_text += f"[{username}]: {text}\n"
            
            # System prompt for Telegram message summarization
            system_prompt = """You are a memory extraction system for Telegram group chats. Your role is to analyze chat messages and extract structured information as JSON.

You MUST return ONLY valid JSON. No markdown code blocks, no explanations, no text before or after the JSON. Start with { and end with }.

Extract and structure the following information as JSON with these EXACT keys:
- summary: A brief 2-3 sentence overview of what was discussed in this chat segment
- keyTopics: Array of main topics discussed (e.g., ["crypto", "AI", "memes"])
- activeUsers: Array of usernames/user_ids that were most active
- sentiment: Overall sentiment of the conversation ("positive", "neutral", "negative", "mixed")
- importantMessages: Array of particularly important or notable messages (max 5)

Example of correct JSON format:
{
  "summary": "The group discussed upcoming NFT drops and debated AI capabilities. Several users shared memes.",
  "keyTopics": ["NFTs", "artificial intelligence", "memes"],
  "activeUsers": ["user123", "cryptofan", "aidev"],
  "sentiment": "positive",
  "importantMessages": ["Check out the new collection dropping tomorrow", "AI is getting scary good"]
}"""
            
            query = f"Summarize these Telegram messages:\n\n{conversation_text}\n\nReturn ONLY the JSON object, nothing else."
            
            result = llm.generate(
                query=query,
                system_prompt=system_prompt,
                quantum_influence=0.2,
                max_length=800,
                conversation_history=None,
                enable_thinking=False
            )
            
            summary_text = result.get('response', '').strip()
            
            # Extract JSON
            from ....utils.json_parser import extract_json_from_llm_response
            summary_data = extract_json_from_llm_response(summary_text, context="telegram_summary")
            return summary_data
            
        except Exception as e:
            logger.error(f"Error summarizing Telegram messages: {e}")
            return None
    
    def execute(self, context: ExecutionContext) -> Dict[str, Any]:
        """Execute telegram memory creator - stores message and potentially summarizes"""
        message = self.get_input_value('message', context, '')
        user_id = self.get_input_value('user_id', context, '')
        username = self.get_input_value('username', context, '')
        chat_id = self.get_input_value('chat_id', context, '')
        storage_instance = self.get_input_value('storage_instance', context, None)
        llm = self.get_input_value('model', context, None)
        
        # Get threshold from metadata
        summarize_threshold = int(self.metadata.get('summarize_threshold', 50))
        if summarize_threshold < 5:
            summarize_threshold = 5
        
        # Validate required inputs
        if not message:
            logger.warning("[TelegramMemoryCreator] No message provided")
            return {'success': False, 'message_count': 0, 'summary_created': False}
        
        if not chat_id:
            logger.warning("[TelegramMemoryCreator] No chat_id provided")
            return {'success': False, 'message_count': 0, 'summary_created': False}
        
        if storage_instance is None:
            raise ValueError("storage_instance is required for TelegramMemoryCreatorNode")
        
        # Create message data
        message_data = {
            'message': str(message),
            'user_id': str(user_id) if user_id else '',
            'username': str(username) if username else '',
            'chat_id': str(chat_id),
            'timestamp': time.time(),
            'type': 'telegram_message'
        }
        
        # Add to buffer for potential summarization
        self._add_to_buffer(storage_instance, str(chat_id), message_data)
        
        # Save individual message to storage
        try:
            # Use activity log to store message
            storage_instance.create_activity_log(
                activity_type='telegram_message',
                message=f'[{username or user_id}] {message[:100]}...' if len(message) > 100 else f'[{username or user_id}] {message}',
                metadata=message_data
            )
            logger.info(f"[TelegramMemoryCreator] Saved message from {username or user_id} in chat {chat_id}: {message[:50]}...")
        except Exception as e:
            logger.error(f"[TelegramMemoryCreator] Failed to save message: {e}")
            return {'success': False, 'message_count': 0, 'summary_created': False}
        
        # Increment count and check threshold
        message_count = self._increment_message_count(storage_instance, str(chat_id))
        should_summarize = message_count > 0 and message_count % summarize_threshold == 0
        summary_created = False
        
        if should_summarize and llm:
            logger.info(f"[TelegramMemoryCreator] Summarizing {summarize_threshold} messages for chat {chat_id}")
            
            # Get messages to summarize
            buffer = self._get_message_buffer(storage_instance, str(chat_id))
            messages_to_summarize = buffer[-summarize_threshold:]
            
            # Create summary
            summary_data = self._summarize_messages(llm, messages_to_summarize, str(chat_id))
            
            if summary_data:
                # Add metadata
                summary_data['chat_id'] = str(chat_id)
                summary_data['message_count'] = summarize_threshold
                summary_data['timestamp'] = time.time()
                
                # Save summary to storage
                storage_instance.create_activity_log(
                    activity_type='telegram_summary',
                    message=f'Chat summary for {chat_id}: {summary_data.get("summary", "")[:100]}...',
                    metadata=summary_data
                )
                
                summary_created = True
                logger.info(f"[TelegramMemoryCreator] Created summary for chat {chat_id}")
            else:
                logger.warning(f"[TelegramMemoryCreator] Failed to create summary for chat {chat_id}")
        
        return {
            'success': True,
            'message_count': message_count,
            'summary_created': summary_created
        }

File: ts/src/core/execution/nodes/telegramMemoryCreator.ts
/**
 * TelegramMemoryCreatorNode – stores Telegram messages and creates summaries per chat.
 * Mirrors Python src/core/execution/nodes/telegram_memory_creator.py
 *
 * Unlike regular MemoryCreator which stores Q&A pairs, this stores
 * individual messages with metadata (user_id, username, chat_id).
 * Tracks message count per chat_id and triggers summarization
 * when threshold is reached.
 *
 * Inputs:
 *   message: Message text (required)
 *   user_id: Telegram user ID (required)
 *   username: Telegram username (optional)
 *   chat_id: Telegram chat/group ID (required)
 *   storage_instance: StorageInterface instance (required)
 *   model: InferenceClient instance (required for summarization)
 *
 * Properties:
 *   summarize_threshold: Number of messages before summarizing (default: 50)
 *
 * Outputs:
 *   success: Boolean indicating if message was stored
 *   message_count: Current message count for this chat
 *   summary_created: True if a summary was just created
 */
import { BaseNode, ExecutionContext } from "../nodeBase";
import { StorageInterface } from "../../types";
import { InferenceClient } from "./inference/inferenceClient";
import { extractJsonFromLlmResponse } from "../../../utils/jsonParser";
import { getLogger } from "../../../utils/logger";

const logger = getLogger("telegramMemoryCreator");

// Class-level caches (shared across all TelegramMemoryCreatorNode instances)
// storage_instance_id → chat_id → count
const messageCounts: Record<string, Record<string, number>> = {};
// storage_instance_id → chat_id → messages[]
const messageBuffers: Record<
  string,
  Record<string, Array<Record<string, unknown>>>
> = {};

/** Unique ID for a storage instance (uses object identity via a WeakMap). */
let storageIdCounter = 0;
const storageIdMap = new WeakMap<object, string>();
function storageId(instance: StorageInterface): string {
  let id = storageIdMap.get(instance);
  if (!id) {
    id = `storage_${storageIdCounter++}`;
    storageIdMap.set(instance, id);
  }
  return id;
}

export class TelegramMemoryCreatorNode extends BaseNode {
  private _summarizeThreshold: number;

  constructor(nodeId: string, nodeData: import("../../types").NodeData) {
    super(nodeId, nodeData);
    this._summarizeThreshold = Number(this.metadata.summarize_threshold ?? 50);
  }

  private _getMessageCount(storage: StorageInterface, chatId: string): number {
    const sid = storageId(storage);
    if (!messageCounts[sid]) messageCounts[sid] = {};
    if (messageCounts[sid][chatId] === undefined) messageCounts[sid][chatId] = 0;
    return messageCounts[sid][chatId];
  }

  private _incrementMessageCount(
    storage: StorageInterface,
    chatId: string
  ): number {
    const sid = storageId(storage);
    if (!messageCounts[sid]) messageCounts[sid] = {};
    if (messageCounts[sid][chatId] === undefined) messageCounts[sid][chatId] = 0;
    messageCounts[sid][chatId]++;
    return messageCounts[sid][chatId];
  }

  private _getMessageBuffer(
    storage: StorageInterface,
    chatId: string
  ): Array<Record<string, unknown>> {
    const sid = storageId(storage);
    if (!messageBuffers[sid]) messageBuffers[sid] = {};
    if (!messageBuffers[sid][chatId]) messageBuffers[sid][chatId] = [];
    return messageBuffers[sid][chatId];
  }

  private _addToBuffer(
    storage: StorageInterface,
    chatId: string,
    messageData: Record<string, unknown>
  ): void {
    const buffer = this._getMessageBuffer(storage, chatId);
    buffer.push(messageData);
    // Keep buffer size reasonable (2x threshold)
    const maxSize = this._summarizeThreshold * 2;
    const sid = storageId(storage);
    if (buffer.length > maxSize) {
      messageBuffers[sid][chatId] = buffer.slice(-maxSize);
    }
  }

  private _clearBuffer(storage: StorageInterface, chatId: string): void {
    const sid = storageId(storage);
    if (messageBuffers[sid]?.[chatId]) {
      messageBuffers[sid][chatId] = [];
    }
  }

  private async _summarizeMessages(
    llm: InferenceClient,
    messages: Array<Record<string, unknown>>
  ): Promise<Record<string, unknown> | null> {
    if (!messages.length) return null;

    try {
      // Format messages for summarization
      let conversationText = "";
      for (const msg of messages) {
        const username =
          (msg.username as string) || (msg.user_id as string) || "Unknown";
        const text = (msg.message as string) || "";
        conversationText += `[${username}]: ${text}\n`;
      }

      const systemPrompt = `You are a memory extraction system for Telegram group chats. Your role is to analyze chat messages and extract structured information as JSON.

You MUST return ONLY valid JSON. No markdown code blocks, no explanations, no text before or after the JSON. Start with { and end with }.

Extract and structure the following information as JSON with these EXACT keys:
- summary: A brief 2-3 sentence overview of what was discussed in this chat segment
- keyTopics: Array of main topics discussed (e.g., ["crypto", "AI", "memes"])
- activeUsers: Array of usernames/user_ids that were most active
- sentiment: Overall sentiment of the conversation ("positive", "neutral", "negative", "mixed")
- importantMessages: Array of particularly important or notable messages (max 5)

Example of correct JSON format:
{
  "summary": "The group discussed upcoming NFT drops and debated AI capabilities. Several users shared memes.",
  "keyTopics": ["NFTs", "artificial intelligence", "memes"],
  "activeUsers": ["user123", "cryptofan", "aidev"],
  "sentiment": "positive",
  "importantMessages": ["Check out the new collection dropping tomorrow", "AI is getting scary good"]
}`;

      const query = `Summarize these Telegram messages:\n\n${conversationText}\n\nReturn ONLY the JSON object, nothing else.`;

      const result = await llm.generate(
        query,
        systemPrompt,
        0.2, // Low quantum_influence for consistent summaries
        800, // Enough tokens for JSON
        null, // No conversation history
        false // No thinking mode
      );

      const summaryText = (result.response ?? "").trim();
      const raw = extractJsonFromLlmResponse(summaryText, "telegram_summary");
      const summaryData = Array.isArray(raw) ? null : (raw as Record<string, unknown>);
      return summaryData;
    } catch (err) {
      logger.error(`Error summarizing Telegram messages: ${err}`);
      return null;
    }
  }

  async execute(context: ExecutionContext): Promise<Record<string, unknown>> {
    const message = this.getInputValue("message", context, "") as string;
    const userId = this.getInputValue("user_id", context, "") as string;
    const username = this.getInputValue("username", context, "") as string;
    const chatId = this.getInputValue("chat_id", context, "") as string;
    const messageIdRaw = this.getInputValue("message_id", context, undefined);
    const num =
      messageIdRaw != null && messageIdRaw !== ""
        ? Number(messageIdRaw)
        : NaN;
    const messageId =
      Number.isFinite(num) && num > 0 && Math.floor(num) === num
        ? num
        : undefined;
    const storage = this.getInputValue(
      "storage_instance",
      context,
      undefined
    ) as StorageInterface | undefined;
    const llm = this.getInputValue("model", context, undefined) as
      | InferenceClient
      | undefined;

    // Normalize message to string immediately
    const messageStr = message ? String(message) : "";

    // Get threshold from metadata and normalize
    let summarizeThreshold = Number(this.metadata.summarize_threshold ?? 50);
    if (summarizeThreshold < 5) summarizeThreshold = 5;
    this._summarizeThreshold = summarizeThreshold;

    // Validate required inputs
    if (!messageStr) {
      logger.warning("[TelegramMemoryCreator] No message provided");
      return { success: false, message_count: 0, summary_created: false };
    }

    const userIdStr = userId ? String(userId) : "";
    if (!userIdStr) {
      logger.warning("[TelegramMemoryCreator] No user_id provided");
      return { success: false, message_count: 0, summary_created: false };
    }

    if (!chatId) {
      logger.warning("[TelegramMemoryCreator] No chat_id provided");
      return { success: false, message_count: 0, summary_created: false };
    }

    if (!storage) {
      throw new Error(
        "storage_instance is required for TelegramMemoryCreatorNode"
      );
    }

    // Create message data (include message_id so storage can resolve message_id → user_id / username later)
    const messageData: Record<string, unknown> = {
      message: messageStr,
      user_id: userId ? String(userId) : "",
      username: username ? String(username) : "",
      chat_id: String(chatId),
      timestamp: Date.now() / 1000,
      type: "telegram_message",
    };
    // Persist message_id so memory selector and TG action can show/resolve it (required for delete/pin/timeout by context)
    if (typeof messageId === "number" && Number.isFinite(messageId)) {
      messageData.message_id = messageId;
    }

    // Save individual message to storage FIRST, then buffer on success
    try {
      const displayMsg =
        messageStr.length > 100 ? messageStr.slice(0, 100) + "..." : messageStr;
      const displayUser = username || userId;
      await storage.createActivityLog("telegram_message", `[${displayUser}] ${displayMsg}`, messageData);
      logger.info(
        `[TelegramMemoryCreator] Saved message from ${displayUser} in chat ${chatId}: ${messageStr.slice(0, 50)}...`
      );
    } catch (err) {
      logger.error(`[TelegramMemoryCreator] Failed to save message: ${err}`);
      return { success: false, message_count: 0, summary_created: false };
    }

    // Add to buffer only after successful persistence
    this._addToBuffer(storage, String(chatId), messageData);

    // Increment count and check threshold
    const messageCount = this._incrementMessageCount(storage, String(chatId));
    const shouldSummarize =
      messageCount > 0 && messageCount % summarizeThreshold === 0;
    let summaryCreated = false;

    if (shouldSummarize && llm) {
      logger.info(
        `[TelegramMemoryCreator] Summarizing ${summarizeThreshold} messages for chat ${chatId}`
      );

      // Get messages to summarize
      const buffer = this._getMessageBuffer(storage, String(chatId));
      const messagesToSummarize = buffer.slice(-summarizeThreshold);

      // Create summary
      const summaryData = await this._summarizeMessages(
        llm,
        messagesToSummarize
      );

      if (summaryData) {
        // Add metadata
        summaryData.chat_id = String(chatId);
        summaryData.message_count = summarizeThreshold;
        summaryData.timestamp = Date.now() / 1000;

        // Save summary to storage (wrapped in try/catch to preserve buffer on failure)
        try {
          const summaryText = (summaryData.summary as string) ?? "";
          const displaySummary =
            summaryText.length > 100
              ? summaryText.slice(0, 100) + "..."
              : summaryText;
          await storage.createActivityLog(
            "telegram_summary",
            `Chat summary for ${chatId}: ${displaySummary}`,
            summaryData
          );

          // Only clear buffer after successful persistence
          this._clearBuffer(storage, String(chatId));

          summaryCreated = true;
          logger.info(
            `[TelegramMemoryCreator] Created summary for chat ${chatId}`
          );
        } catch (err) {
          // Don't clear buffer so it's available for retry on next threshold hit
          logger.error(
            `[TelegramMemoryCreator] Failed to persist summary for chat ${chatId}: ${err}`
          );
        }
      } else {
        logger.warning(
          `[TelegramMemoryCreator] Failed to create summary for chat ${chatId}`
        );
      }
    }

    return {
      success: true,
      message_count: messageCount,
      summary_created: summaryCreated,
    };
  }
}
